### PR TITLE
fixes to ORSet mergeRemoveDelta and ORMap deltaMerge (#22648)

### PR DESCRIPTION
**Pull request by gosubpl**
* fix issue with ORMultiMap remove #22648
* fix issue with Failed: ReplicatorMapDeltaSpec #22631
* fix ORSet.mergeRemoveDelta
* fix ORMap.mergeDelta
* add detailed test for ORMultiMap.withValueDeltas

### CODE REVIEW

File: akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMultiMapSpec.scala [hunk 1]
```diff
@@ -89,6 +86,47 @@ class ORMultiMapSpec extends WordSpec with Matchers {
     }
   }
 
+  "be able to have its entries correctly merged with another ORMultiMap with overlapping entries 2" in {
```
Comment (gosubpl) [inline akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMultiMapSpec.scala]: Test from #22648

File: akka-distributed-data/src/test/scala/akka/cluster/ddata/ORSetSpec.scala [hunk 2]
```diff
@@ -434,6 +419,40 @@ class ORSetSpec extends WordSpec with Matchers {
       t4.elements should ===(Set("b", "a"))
     }
 
+    "not have anomalies for ORSet in complex but realistic scenario" in {
```
Comment (gosubpl) [inline akka-distributed-data/src/test/scala/akka/cluster/ddata/ORSetSpec.scala]: Test from #22631

**Comment by gosubpl** [comment]
Supersedes https://github.com/akka/akka/pull/22654
Fixes #22631 and #22648
Tested on repeat with `ReplicatorMapDeltaSpec` 200 iterations with no failure. Also all test cases reproduced from failing tests instances were tested.

**Comment by gosubpl** [comment]
Contains brand new implementation of `mergeDelta` for `ORMap`, plus the already discussed fix for `ORSet` `mergeRemoveDelta` op.
Also, in one of the tests there is analysis of `ORMultiMap.withValuesDelta`, showing that the operations from the public API of this type can only create _trivial_ tombstones (i.e. equal to `ORSet[B]`), which opens a way to do away with those tombstones completely.

**Comment by gosubpl** [comment]
I have run a couple of hundred iterations of `ReplicatedMapDeltaSpec` without issues. Also, some 20 iterations of this spec for 3600 random ops each.

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORSet.scala [hunk 3]
```diff
@@ -440,7 +438,7 @@ final class ORSet[A] private[akka] (
     def deleteDots = that.vvector.versionsIterator
     def deleteDotsNodes = deleteDots.map { case (dotNode, _) ⇒ dotNode }
     val newElementsMap =
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORSet.scala]: Slightly more efficient and perhaps also more readable?

```scala
    val newElementsMap = {
      val thisDotOption = this.elementsMap.get(elem)
      val deleteDotsAreGreater = deleteDots.forall {
        case (dotNode, dotV) ⇒
          thisDotOption match {
            case Some(thisDot) ⇒ thisDot.versionAt(dotNode) <= dotV
            case None          ⇒ false
          }
      }
      if (deleteDotsAreGreater) {
        thisDotOption match {
          case Some(thisDot) ⇒
            if (thisDot.versionsIterator.forall { case (thisDotNode, _) ⇒ deleteDotsNodes.contains(thisDotNode) })
              elementsMap - elem
            else elementsMap
          case None ⇒
            elementsMap
        }
      } else
        elementsMap
    }
```

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala [hunk 4]
```diff
@@ -97,24 +129,44 @@ object ReplicatorMapDeltaSpec extends MultiNodeConfig {
         availableForRemove.toVector(rnd.nextInt(availableForRemove.size))
     }
 
-    (0 to (30 + rnd.nextInt(10))).map { _ ⇒
-      rnd.nextInt(4) match {
+    (0 to (120 + rnd.nextInt(10))).map { _ ⇒
```
Comment (patriknw) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala]: I'm not sure it is better coverage just by increasing this, e.g. higher chance that a Set will just be filled with all values if removals are not used. 50?

**Review by patriknw — APPROVED** [review]
LGTM, with optional comments

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORSet.scala (hunk 3 above)
Comment (gosubpl) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORSet.scala]: perhaps. thanks!

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala (hunk 4 above)
Comment (gosubpl) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala]: yes, that could be the case. But with 50 still passes repeat locally.

Comment (gosubpl) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala]: the trade-off is that with a very small number removals are generated very infrequently, which was the cause of not getting failing case in repeat with previous version of `mergeDelta`

**Comment by gosubpl** [comment]
@patriknw I've included your comments. Do you think it would be good to get this on repeat for a while before merging?

**Comment by patriknw** [comment]
I have been running all tests in jenkins repeat during weekend. We can merge tomorrow. Thanks for great work.

**Comment by patriknw** [comment]
I trust the thorough testing here

**Comment by raboof** [comment]
Cool fixes!

**Comment by gosubpl** [comment]
@raboof thanks!
@patriknw Many thanks for all the help, assistance & patience! That might not provide too much comfort for you, but I always stand behind the code I write or touch :)

**Comment by patriknw** [comment]
@gosubpl thanks again for great work

**Pull request merged**